subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,230 @@
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
+		return;
+	} else if (len <= 0) {
+		return;
+	}
+
+	if (state == ICMSG_STATE_READY) {
+		if (dev_data->ops && dev_data->ops->received) {
+			dev_data->ops->received(cb_buffer, len, NULL);
+		} else {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		/* Reading 0 to know if there are data in the buffer. */
+		len = icmsg_buf_read(dev_data->rx_ib, NULL, 0);
+		if (len > 0) {
+			(void)k_work_submit(&dev_data->mbox_work);
+		}
+	} else {
+		__ASSERT_NO_MSG(state == ICMSG_STATE_BUSY);
+		if (len == sizeof(magic) && !memcmp(magic, cb_buffer, len)) {
+			if (dev_data->ops && dev_data->ops->bound) {
+				dev_data->ops->bound(NULL);
+			} else {
+				__ASSERT_NO_MSG(false);
+				return;
+			}
+
+			atomic_set(&dev_data->state, ICMSG_STATE_READY);
+		} else {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+	}
+}
+
+static void mbox_callback(const struct device *instance, uint32_t channel,
+			  void *user_data, struct mbox_msg *msg_data)
+{
+	struct backend_data_t *dev_data = user_data;
+
+	(void)k_work_submit(&dev_data->mbox_work);
+}
+
+static int mbox_init(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	int err;
+
+	k_work_init(&dev_data->mbox_work, mbox_callback_process);
+
+	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, dev_data);
+	if (err != 0) {
+		return err;
+	}
+
+	return mbox_set_enabled(&conf->mbox_rx, 1);
+}
+
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	struct ipc_ept *ep = (struct ipc_ept *)token;
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_BUSY) {
+		/* Endpoint is still being initialized. */
+		return -EBUSY;
+	}
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_READY) {
+		/* Already registered. This backend support single ep. */
+		return -EALREADY;
+	}
+
+	ep->instance = instance;
+	dev_data->ops = &cfg->cb;
+
+	atomic_set(&dev_data->state, ICMSG_STATE_BUSY);
+
+	int ret = mbox_init(instance);
+
+	if (ret) {
+		return ret;
+	}
+
+	ret = icmsg_buf_write(dev_data->tx_ib, magic, sizeof(magic));
+	if (ret < sizeof(magic)) {
+		__ASSERT_NO_MSG(ret == sizeof(magic));
+		return ret;
+	}
+
+	ret = mbox_send(&conf->mbox_tx, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	(void)k_work_submit(&dev_data->mbox_work);
+
+	return 0;
+}
+
+static int send(const struct device *instance, void *token,
+		const void *msg, size_t len)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+
+	if (atomic_get(&dev_data->state) != ICMSG_STATE_READY) {
+		return -EBUSY;
+	}
+
+	/* Empty message is not allowed */
+	if (len == 0) {
+		return -ENODATA;
+	}
+
+	int ret = icmsg_buf_write(dev_data->tx_ib, msg, len);
+
```

Comment:
empty line

Reply:
I can't remove this empty line. This style is according to coding style. Checkpatch will complain if i remove it.